torchtune/data/_collate.py

```diff
@@ -12,7 +12,136 @@
+    batch of sequences with the given `pad_fn` to the maximum sequence length for
+    each entry in the batch.
+
+    Note:
+        This function assumes all batch elements which are not in ``keys_to_pad`` do not require
+        any collation (see example below).
+
+    Args:
+        batch (List[Dict[str, List[int]]]): A list of dictionaries containing inputs.
+        pad_fn (Callable): padding function to apply to ``keys_to_pad`` batch elements.
+            This should have an identical signature to :func:`torch.nn.utils.rnn.pad_sequence`, e.g.
+            :func:`torchtune.data.left_pad_sequence`.
+        keys_to_pad (List[str]): Batch element keys to apply padding to. Should be a subset
+            of keys in the batch.
+        padding_idx (Union[int, Dict[str, int]]): Either a single integer padding value to apply to all
+            ``keys_to_pad`` elements, or a mapping with keys identical to ``keys_to_pad`` with per-key
+            padding values.
+
+    Returns:
+        torch.Tensor: The padded tensor of input ids with shape [batch_size, max_seq_len].
+
+    Raises:
+        ValueError: if ``keys_to_pad`` is empty, or is not a list, or is not a subset of keys in the batch.
+        ValueError: if ``padding_idx`` is provided as a dictionary, but the keys are not identical to
+            ``keys_to_pad``.
+
+    Example:
+        >>> a = [1, 2, 3]
+        >>> b = [4, 5, 6, 7]
+        >>> c = [8, 9, 10, 11, 12]
+        >>> batch = [
+        >>>     {"tokens": a, "labels": 1},
+        >>>     {"tokens": b, "labels": 3},
+        >>>     {"tokens": c, "labels": 0},
+        >>> ]
+        >>> padded_collate(
+        >>>     batch,
+        >>>     pad_fn=left_pad_sequence,
+        >>>     keys_to_pad=["tokens"],
+        >>>     padding_idx=-10
+        >>> )
+        {
+            'labels': tensor([1, 3, 0]),
+            'tokens': tensor([[-10, -10,   1,   2,   3],
+                              [-10,   4,   5,   6,   7],
+                              [  8,   9,  10,  11,  12]])
+        }
+    """
+    if not isinstance(keys_to_pad, list) or not keys_to_pad:
+        raise ValueError(
+            f"keys_to_pad should be a list of strings with at least one element, but found {keys_to_pad}!"
+        )
+    if isinstance(padding_idx, dict):
+        if isinstance(padding_idx, dict) and not set(padding_idx.keys()) == set(
+            keys_to_pad
+        ):
+            raise ValueError(
+                f"padding_idx was provided as a dictionary, but the keys ({padding_idx.keys()}) "
+                f"are not the same as keys_to_pad ({keys_to_pad})"
+            )
+        if not set(keys_to_pad) <= set(batch[0].keys()):
+            raise ValueError(
+                "keys_to_pad should be a subset of keys in the batch, but found "
+                f"{set(keys_to_pad)} and {set(batch[0].keys())}, respectively."
+            )
+
+    batch_keys = [k for k in batch[0].keys() if k not in keys_to_pad]
```

Comment:
a quick comment explaining that this is taking out keys that won't be padded would be good